diff --git a/Android.mk b/Android.mk
new file mode 100644
index 0000000..b4324fd
--- /dev/null
+++ b/Android.mk
@@ -0,0 +1,15 @@
+LOCAL_PATH := $(call my-dir)
+
+include $(CLEAR_VARS)
+LOCAL_CFLAGS    := $(APP_CFLAGS)
+LOCAL_CPPFLAGS  := $(APP_CPPFLAGS)
+LOCAL_ARM_MODE  := $(APP_ARM_MODE)
+
+LOCAL_MODULE    := hqx
+
+LOCAL_SRC_FILES := \
+	src/hq4x.c \
+	src/hq2x.c \
+	src/hq3x.c
+
+include $(BUILD_STATIC_LIBRARY)
diff --git a/src/hq2x.c b/src/hq2x.c
index 14740b1..39b3817 100644
--- a/src/hq2x.c
+++ b/src/hq2x.c
@@ -19,7 +19,7 @@
  */
 
 #include <stdint.h>
-#include "common.h"
+#include "hqxcommon.h"
 #include "hqx.h"
 
 #define PIXEL00_0     *dp = w[5];
diff --git a/src/hq3x.c b/src/hq3x.c
index bf5588a..c07ac63 100644
--- a/src/hq3x.c
+++ b/src/hq3x.c
@@ -19,7 +19,7 @@
  */
 
 #include <stdint.h>
-#include "common.h"
+#include "hqxcommon.h"
 #include "hqx.h"
 
 #define PIXEL00_1M  Interp1(dp, w[5], w[1]);
diff --git a/src/hq4x.c b/src/hq4x.c
index e7cf57a..230f662 100644
--- a/src/hq4x.c
+++ b/src/hq4x.c
@@ -19,7 +19,7 @@
  */
 
 #include <stdint.h>
-#include "common.h"
+#include "hqxcommon.h"
 #include "hqx.h"
 
 #define PIXEL00_0     *dp = w[5];
diff --git a/src/common.h b/src/hqxcommon.h
similarity index 93%
rename from src/common.h
rename to src/hqxcommon.h
index 2f56953..c13e708 100644
--- a/src/common.h
+++ b/src/hqxcommon.h
@@ -37,13 +37,18 @@
 #define trU   0x00000700
 #define trV   0x00000006
 
-/* RGB to YUV lookup table */
-extern uint32_t RGBtoYUV[16777216];
-
 static inline uint32_t rgb_to_yuv(uint32_t c)
 {
-    // Mask against MASK_RGB to discard the alpha channel
-    return RGBtoYUV[MASK_RGB & c];
+	int r, g, b, y, u, v;
+
+    r = (c & 0xFF0000) >> 16;
+    g = (c & 0x00FF00) >> 8;
+    b = c & 0x0000FF;
+    y = (uint32_t)(0.299*r + 0.587*g + 0.114*b);
+    u = (uint32_t)(-0.169*r - 0.331*g + 0.5*b) + 128;
+    v = (uint32_t)(0.5*r - 0.419*g - 0.081*b) + 128;
+
+    return (y << 16) + (u << 8) + v;
 }
 
 /* Test if there is difference in color */
